charms/worker/k8s/src/charm.py

```diff
@@ -232,6 +250,27 @@ def _configure_cos_integration(self):
         if relation := self.model.get_relation("cos-tokens"):
             self.collector.request(relation)
 
+    def _configure_etcd(self, config: BootstrapConfig):
+        """Configure etcd as the datastore for the Kubernetes cluster.
+
+        Args:
+            config (BootstrapConfig): The bootstrap configuration object for
+                the Kubernetes cluster that is being configured. This object
+                will be modified in-place to include etcd's configuration details.
+        """
+        etcd_relation = self.model.get_relation("etcd")
+        if not etcd_relation:
+            return
```

Comment:
During the test runs, I have noticed that the leader waits for the etcd cluster to become available. I wonder if this is a change in the current ops. It seems that the relation object is present in the environment, but with no units in the relation. I think if we have a relation to etcd, but no units yet, the object will be present but with an unready relation.

Reply:
you're right that it just needs "etcd-relation-created" for the relation to exist.  I'm just afraid that hook doesn't come before the cluster is bootstrapped.  Something has me concerned the cluster would bootstrap before this hook and the relation wouldn't have been created yet.

Reply:
> you're right that it just needs etcd-relation-created for the relation to exist. I'm just afraid that hook doesn't come before the cluster is bootstrapped.

Right, juju will first do a cycle of `install`, `config-changed` hooks, then fire the `*-relation-created` events. In this scenario, during the initial `install` hook we cannot know if there's an etcd relation waiting to be established, therefore our easiest way out (at the moment) is a config option.

To be fair, I have not tested this with more recent juju versions, so this might have been resolved. However:
- we should not rely on this behaviour if it's not explicitly part of the juju "contract". Even if the _timings_ are such that _99.999%_ percent it goes like we want it to, it would hurt as a lot to design such a race condition that leads to DoA clusters (or, even worse, clusters with an unpredictable state)
- even if it becomes standard juju behaviour in the future, i would still be wary to use it because it would not work with older juju versions, and that's a very dangerous path to tread on.

I think it's preferable for us to be able to control that the charm behaves _exactly_ the way that we want it to.

Reply:
Maybe charm config like this?  the charm can be deployed via bundle or cli -- and in both cases the config is available on the install hook

```suggestion
        etcd_external = self.config.get("etcd-external")
        if not etcd_external:
            log.info("Charm config allows for bootstrap without an external etcd")
            return
        etcd_relation = self.model.get_relation("etcd")
```

Reply:
Thank you for your comments, @addyess and @neoaggelos. I have introduced a new configuration option that should resolve this issue. I have opted for a configuration option for the name of the `datastore`, with `dqlite` being the default one.